From 4421eb28952cc3ff61c77ff331e0e7c77e587b4a Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Fri, 17 Sep 2021 10:32:33 +0200
Subject: [PATCH 31/73] boot: use cleanup-based file handle closing a bit more

---
 src/boot/efi/boot.c | 8 ++------
 1 file changed, 2 insertions(+), 6 deletions(-)

diff --git a/src/boot/efi/boot.c b/src/boot/efi/boot.c
index cfe20c27f..106fda925 100644
--- a/src/boot/efi/boot.c
+++ b/src/boot/efi/boot.c
@@ -1492,7 +1492,7 @@ static VOID config_load_entries(
                 EFI_FILE *root_dir,
                 CHAR16 *loaded_image_path) {
 
-        EFI_FILE_HANDLE entries_dir;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE entries_dir = NULL;
         EFI_STATUS err;
 
         assert(config);
@@ -1530,8 +1530,6 @@ static VOID config_load_entries(
                 if (!EFI_ERROR(err))
                         config_entry_add_from_file(config, device, root_dir, L"\\loader\\entries", f->FileName, content, loaded_image_path);
         }
-
-        uefi_call_wrapper(entries_dir->Close, 1, entries_dir);
 }
 
 static INTN config_entry_compare(ConfigEntry *a, ConfigEntry *b) {
@@ -1982,7 +1980,7 @@ static VOID config_entry_add_linux(
                 EFI_HANDLE *device,
                 EFI_FILE *root_dir) {
 
-        EFI_FILE_HANDLE linux_dir;
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE linux_dir = NULL;
         EFI_STATUS err;
         ConfigEntry *entry;
 
@@ -2112,8 +2110,6 @@ static VOID config_entry_add_linux(
                 FreePool(os_build_id);
                 FreePool(content);
         }
-
-        uefi_call_wrapper(linux_dir->Close, 1, linux_dir);
 }
 
 #define XBOOTLDR_GUID \
-- 
2.34.1

